rollup/sync_service.go

```diff
@@ -0,0 +1,844 @@
+	event, err := s.ctcFilterer.ParseSequencerBatchAppended(ethlog)
+	if err != nil {
+		return err
+	}
+
+	tx, pending, err := s.ethclient.TransactionByHash(ctx, ethlog.TxHash)
+	if err == ethereum.NotFound {
+		log.Error("Transaction not found", "hash", ethlog.TxHash.Hex())
+		return err
+	}
+	if err != nil {
+		log.Error("Cannot fetch transaction", "hash", ethlog.TxHash.Hex())
+		return err
+	}
+	if pending {
+		log.Error("Transaction unexpectedly in mempool", "hash", ethlog.TxHash.Hex())
+		return err
+	}
+
+	cd := appendSequencerBatchCallData{}
+	err = cd.Decode(bytes.NewReader(tx.Data()))
+	if err != nil {
+		return err
+	}
+
+	// event.StartingQueueIndex
+	// event.NumQueueElements
+
+	// event.TotalElements
+	for i, element := range cd.ChainElements {
+		var tx *types.Transaction
+		index := event.TotalElements.Uint64() - (uint64(i) + event.NumQueueElements.Uint64())
+		// Sequencer transaction
+		if element.IsSequenced {
+			nonce := uint64(0)
+			to := common.Address{} // sequencerDecompressionAddress
+			gasLimit := uint64(0)  // max gas limit
+			l1TxOrigin := common.Address{}
+			tx = types.NewTransaction(nonce, to, big.NewInt(0), gasLimit, big.NewInt(0), element.TxData, &l1TxOrigin, nil, types.QueueOriginSequencer, types.SighashEIP155)
+			// Check if there is already a transaction at the index
+			if block := s.bc.GetBlockByNumber(index); block != nil {
+				// Check if the transaction is different
+				// this check will not work, need to check the items in the tx
+				// instead.
+				if included := block.Transaction(tx.Hash()); included == nil {
+					previous := block.Transactions()[0]
+					log.Info("Different transaction detected, reorganizing", "new", tx.Hash().Hex(), "previous", previous.Hash().Hex())
+					err := s.bc.SetHead(index - 1)
+					if err != nil {
+						log.Error("Cannot reorganize")
+						return err
+					}
+				}
+			}
+		} else {
+			// Queue transaction
+			result, ok := s.l1ToL2Enqueued.Load(index)
+			if !ok {
+				log.Error("Cannot find transaction in transaction cache", "index", index)
+				continue
+			}
+			rtx, ok := result.(RollupTransaction)
+			if !ok {
+				log.Error("Incorrect type in transaction cache")
+				continue
+			}
+			tx = rtx.tx
+		}
+
+		s.applyTransaction(tx)
+	}
+
+	return nil
+}
+
+func (s *SyncService) ProcessQueueBatchAppendedLog(ctx context.Context, ethlog types.Log) error {
+	event, err := s.ctcFilterer.ParseQueueBatchAppended(ethlog)
+	if err != nil {
+		return err
+	}
+
+	start := event.StartingQueueIndex.Uint64()
+	end := start + event.NumQueueElements.Uint64()
+
+	for i := start; i < end; i++ {
+		result, ok := s.l1ToL2Enqueued.Load(i)
+		if !ok {
+			// This is a very bad error, how to recover?
+			log.Error("Unknown transaction")
+			continue
+		}
+		rtx, ok := result.(RollupTransaction)
+		if !ok {
+			log.Error("Incorrect type in map")
+			continue
+		}
+
+		tx := rtx.tx
```

Comment:
Add reorg logic that is in the sequencer batch appended function here. If we get a queue tx that isn't equal to the queue tx in the chain, we gotta reorg!